src/ripple/shamap/SHAMapTreeNode.h

```diff
@@ -51,112 +51,163 @@ class SHAMapTreeNode
     bool isLeaf () const;
     bool isInner () const;
-    bool isInBounds (SHAMapNodeID const &id) const;
     bool isValid () const;
+    bool isInBounds (SHAMapNodeID const &id) const;
+
+    virtual bool updateHash () = 0;
+    virtual void addRaw (Serializer&, SHANodeFormat format) = 0;
+    virtual std::string getString (SHAMapNodeID const&) const;
+    virtual std::shared_ptr<SHAMapAbstractNode> clone(std::uint32_t seq) const = 0;
+
+    static std::shared_ptr<SHAMapAbstractNode>
+        make(Blob const& rawNode, std::uint32_t seq, SHANodeFormat format,
+             uint256 const& hash, bool hashValid);
+
+    // debugging
+#ifdef BEAST_DEBUG
+    static void dump (SHAMapNodeID const&, beast::Journal journal);
+#endif
+};
+
+class SHAMapInnerNode
+    : public SHAMapAbstractNode
+{
+    uint256                         mHashes[16];
+    std::shared_ptr<SHAMapAbstractNode> mChildren[16];
+    int                             mIsBranch = 0;
+    std::uint32_t                   mFullBelowGen = 0;
+
+    static std::mutex               childLock;
+public:
+    SHAMapInnerNode(std::uint32_t seq = 0);
+    std::shared_ptr<SHAMapAbstractNode> clone(std::uint32_t seq) const override;
 
-    // inner node functions
-    bool isInnerNode () const;
-    bool isEmptyBranch (int m) const;
     bool isEmpty () const;
+    bool isEmptyBranch (int m) const;
     int getBranchCount () const;
-    void makeInner ();
     uint256 const& getChildHash (int m) const;
 
-    // item node function
-    bool hasItem () const;
-    std::shared_ptr<SHAMapItem> const& peekItem () const;
-    bool setItem (std::shared_ptr<SHAMapItem> const& i, TNType type);
+    void setChild(int m, std::shared_ptr<SHAMapAbstractNode> const& child);
+    void shareChild (int m, std::shared_ptr<SHAMapAbstractNode> const& child);
+    SHAMapAbstractNode* getChildPointer (int branch);
+    std::shared_ptr<SHAMapAbstractNode> getChild (int branch);
+    std::shared_ptr<SHAMapAbstractNode>
+        canonicalizeChild (int branch, std::shared_ptr<SHAMapAbstractNode> node);
 
     // sync functions
     bool isFullBelow (std::uint32_t generation) const;
     void setFullBelowGen (std::uint32_t gen);
 
-    SHAMapTreeNode* getChildPointer (int branch);
-    std::shared_ptr<SHAMapTreeNode> getChild (int branch);
-    void canonicalizeChild (int branch, std::shared_ptr<SHAMapTreeNode>& node);
-
-    // debugging
-#ifdef BEAST_DEBUG
-    void dump (SHAMapNodeID const&, beast::Journal journal);
-#endif
-    std::string getString (SHAMapNodeID const&) const;
-    bool updateHash ();
+    bool updateHash () override;
     void updateHashDeep();
+    void addRaw (Serializer&, SHANodeFormat format) override;
+    std::string getString (SHAMapNodeID const&) const override;
+
+    friend std::shared_ptr<SHAMapAbstractNode>
+        SHAMapAbstractNode::make(Blob const& rawNode, std::uint32_t seq,
+             SHANodeFormat format, uint256 const& hash, bool hashValid);
+};
 
+class SHAMapTreeNode
```

Comment:
In the interest of reducing changes to make this thing reviewable, I recycled the existing name `SHAMapTreeNode` to mean "leaf node" since this is what most clients of `SHAMap` traffic in.  A future refactor can more safely rename this class to something else if desired.